Fix reassigning to class name during static field initialization

hyjorc1 opened the pull request:
#### 68f502ba813d3162e60f92023084d72c167bbc7f
<pre>
Fix reassigning to class name during static field initialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=247428">https://bugs.webkit.org/show_bug.cgi?id=247428</a>
rdar://102193862

Reviewed by Alexey Shvayka.

Our investigation shows that there is a different path for `{ x = 0 } = 0`
by comparing to `{ x } = 0`, which go through `parseMemberExpression`
=&gt; `parsePrimaryExpression` =&gt; `createResolveAndUseVariable` =&gt;
`currentScope()-&gt;useVariable()`. `useVariable` is necessary so the
BytecodeGenerator would know it&apos;s used via scope VarKind::Scope and
set it up via op_put_to_scope rather than simple mov, which used to
trigger TDZ error.

For `{ x = 0 } = 0`, the parser will go through `parseProperty` and step
into the branch
```
if (match(EQUAL))
    classifyExpressionError(ErrorIndicatesPattern);
```
which fails to updating the used variables when encountering the
object destructuring with assignment expressions. This fix is to
handle those edge cases. In this case, symbolTablePut will check
for read-only variable without ignoring.

* JSTests/stress/destructuring-array-class-static.js: Added.
(shouldThrow):
(x):
* JSTests/stress/destructuring-object-class-static.js: Added.
(shouldThrow):
(x):
(eval.x):
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser&lt;LexerType&gt;::parseDestructuringPattern):

Canonical link: <a href="https://commits.webkit.org/256663@main">https://commits.webkit.org/256663@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c9a75b7de43874230b611832c3d8175b6d2d746

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96452 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5706 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29523 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105995 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166347 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5886 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34452 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88831 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102720 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102124 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4390 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83056 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31366 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74265 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/87436 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40182 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/82833 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37855 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21002 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28303 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4621 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4130 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/85511 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/938 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40272 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19275 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

shvaikalesh reviewed (approved):
I've investigated the issue from beginning and ended up with the same patch, but with a bit different explanation:

We have a different (fast?) path for things like `let { x = 1 } = {}`, which unlike other destructurings doesn't take `parseMemberExpression` => `parsePrimaryExpression` => `createResolveAndUseVariable` => `currentScope()->useVariable()`, so it's necessary to at least `useVariable(<className>)` as well.

We need to `useVariable(<className>)` so the BytecodeGenerator would know it's used via scope `VarKind::Scope` and set it up via `op_put_to_scope` rather than simple `mov`, which used to trigger TDZ error.

shvaikalesh commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -1297,6 +1297,8 @@ template <class TreeBuilder> TreeDestructuringPattern Parser<LexerType>::parseDe
                     }
                     semanticFailIfTrue(isDisallowedIdentifierAwait(identifierToken), "Cannot use 'await' as a ", destructuringKindToVariableKindName(kind), " ", disallowedIdentifierAwaitReason());
+                    if (kind == DestructuringKind::DestructureToExpressions && match(EQUAL))
+                        currentScope()->useVariable(propertyName, m_vm.propertyNames->eval == *propertyName);
```
Nit: we could hoist

```cpp
semanticFailIfTrue(isDisallowedIdentifierAwait(identifierToken), "Cannot use 'await' as a ", destructuringKindToVariableKindName(kind), " ", disallowedIdentifierAwaitReason());
```

before the branch, right next to `escapedKeyword` check which looks nicer, and also end up dropping extra `kind == DestructuringKind::DestructureToExpressions` check:

```cpp
else {
    semanticFailIfTrue(escapedKeyword, "Cannot use abbreviated destructuring syntax for keyword '", propertyName->impl(), "'");
    semanticFailIfTrue(isDisallowedIdentifierAwait(identifierToken), "Cannot use 'await' as a ", destructuringKindToVariableKindName(kind), " ", disallowedIdentifierAwaitReason());
    if (kind == DestructuringKind::DestructureToExpressions) {
        bool isEvalOrArguments = m_vm.propertyNames->eval == *propertyName || m_vm.propertyNames->arguments == *propertyName;
        if (isEvalOrArguments && strictMode())
            reclassifyExpressionError(ErrorIndicatesPattern, ErrorIndicatesNothing);
        failIfTrueIfStrict(isEvalOrArguments, "Cannot modify '", propertyName->impl(), "' in strict mode");
        if (match(EQUAL))
            currentScope()->useVariable(propertyName, m_vm.propertyNames->eval == *propertyName);
    }
    innerPattern = createBindingPattern(context, kind, exportType, *propertyName, identifierToken, bindingContext, duplicateIdentifier);
}
```

hyjorc1 committed:
Fix reassigning to class name during static field initialization
https://bugs.webkit.org/show_bug.cgi?id=247428
rdar://102193862

Reviewed by Alexey Shvayka.

Our investigation shows that there is a different path for `{ x = 0 } = 0`
by comparing to `{ x } = 0`, which go through `parseMemberExpression`
=> `parsePrimaryExpression` => `createResolveAndUseVariable` =>
`currentScope()->useVariable()`. `useVariable` is necessary so the
BytecodeGenerator would know it's used via scope VarKind::Scope and
set it up via op_put_to_scope rather than simple mov, which used to
trigger TDZ error.

For `{ x = 0 } = 0`, the parser will go through `parseProperty` and step
into the branch
```
if (match(EQUAL))
    classifyExpressionError(ErrorIndicatesPattern);
```
which fails to updating the used variables when encountering the
object destructuring with assignment expressions. This fix is to
handle those edge cases. In this case, symbolTablePut will check
for read-only variable without ignoring.

* JSTests/stress/destructuring-array-class-static.js: Added.
(shouldThrow):
(x):
* JSTests/stress/destructuring-object-class-static.js: Added.
(shouldThrow):
(x):
(eval.x):
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser<LexerType>::parseDestructuringPattern):

Canonical link: https://commits.webkit.org/256663@main
